scheduler/state_machine/state_machine_utils.py

```diff
@@ -821,6 +820,60 @@ def run_uncompleted_dialog(user_id: int):
     else:
         run_option_menu(user_id)
 
+def select_dialog_to_complete(uncompleted_dialogs: List[UserInterventionState], dialog_preference: Components) -> \
+        Optional[UserInterventionState]:
+    """
+    Selects the dialog from the list of uncompleted dialogs for the user to go through with. If there is no preference,
+    then we select a random dialog which isn't a general activity dialog. If all else fails we take the first one.
+
+    Args:
+        uncompleted_dialogs: A list of dialogs that are marked as uncompleted in the database
+        dialog_preference: A specific dialog that takes priority for selection
+    """
+    if len(uncompleted_dialogs) == 0:
+        return None
+
+    general_activity_id = get_component_id(Components.GENERAL_ACTIVITY)
+
+    if dialog_preference is not None:
+        preference_id = get_component_id(dialog_preference)
+        for dialog in uncompleted_dialogs:
+            if dialog.intervention_component_id == preference_id:
+                return dialog
+    else:
+        for dialog in uncompleted_dialogs:
+            if dialog.intervention_component_id != general_activity_id:
+                return dialog
+
+    return uncompleted_dialogs[0]
+
+def get_component_id(dialog: Components) -> Optional[int]:
+    """
+    For a given dialog, retrieve its ID from the database.
+
+    Args:
+        dialog: The component for which to retrieve the ID
+    """
+    session = get_db_session()
+
+    intervention_component = (
+        session.query(
+            InterventionComponents
+        )
+        .filter(
+            InterventionComponents.intervention_component_name == dialog.value
+        )
+        .first()
+    )
+
+    session.close()
+
+    if intervention_component is None:
+        logging.error(dialog.value, " not found in the database")
+        return None
+
+    return intervention_component.intervention_component_id

```

Comment:
When the session is closed you have no more access to the query result, so you have to store the value you want to return in a new variable.

```suggestion
    id = intervention_component.intervention_component_id
    session.close()

    if intervention_component is None:
        logging.error(dialog.value, " not found in the database")
        return None

    return id

```